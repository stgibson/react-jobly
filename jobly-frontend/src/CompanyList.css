.CompanyList a {
  color: black;
}

/* Learned how to style hovering over link at https://www.w3schools.com/cssref/sel_hover.asp */
.CompanyList a:hover {
  color: blue;
  text-decoration: none;
}