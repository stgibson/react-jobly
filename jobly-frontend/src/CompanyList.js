import React, { useState } from "react";
import Container from "react-bootstrap/Container";
import Form from "react-bootstrap/Form";
import Row from "react-bootstrap/Row";
import Col from "react-bootstrap/Col";
import Button from "react-bootstrap/Button";
import CompanyCard from "./CompanyCard";

/**
 * Component for showing list of companies and search box for filtering
 * companies
 * @param {Object{Object{string|number}|function}} param0 
 * @returns JSX code for rendering company list
 */
const CompanyList = ({ companies, findAllCompanies }) => {
  const [filter, setFilter] = useState("");

  /**
   * Updates filter when user types in searchbox
   * @param {Object} evt 
   */
  const handleChange = evt => {
    const { value } = evt.target;
    setFilter(value);
  };

  /**
   * Stops reloading page and gets filtered companies
   * @param {Object} evt 
   */
  const handleSubmit = evt => {
    evt.preventDefault();
    if (!filter) {
      findAllCompanies();
    }
    else {
      findAllCompanies(filter);
    }
  };

  // Learned how to use react-bootstrap forms at https://react-bootstrap.github.io/components/forms/
  // Learned how to use react-bootstrap container at https://react-bootstrap.github.io/layout/grid/
  // Learned how to change padding at https://mdbootstrap.com/docs/react/utilities/spacing/
  return (
    <Container>
      <Form onSubmit={ handleSubmit }>
        <Form.Group as={ Row } controlId="formSearchBox">
          <Col xs={ { span: 6, offset: 2 } } className="p-0">
            <Form.Control type="text" placeholder="Enter search term..." onChange={ handleChange } value={ filter } />
          </Col>
          <Col xs={ 2 }>
            <Button variant="primary" type="submit">Submit</Button>
          </Col>
        </Form.Group>
      </Form>
      <Row>
        <Col xs={ { span: 8, offset: 2 } } className="p-0">
          { companies.map(company => <CompanyCard key={ company.handle } company={ company } />) }
        </Col>
      </Row>
    </Container>
  );
};

export default CompanyList;