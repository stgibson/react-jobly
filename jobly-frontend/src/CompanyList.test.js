import React from "react";
import { MemoryRouter } from "react-router-dom";
import { render, fireEvent } from "@testing-library/react";
import CompanyList from "./CompanyList";

const companies = [
  {
    handle: "test-1",
    name: "Test 1",
    description: "Test Description 1",
    numEmployees: 100,
    logoUrl: "/logos/logo1.png"
  },
  {
    handle: "test-2",
    name: "Test 2",
    description: "Test Description 2",
    numEmployees: 200,
    logoUrl: "/logos/logo2.png"
  },
  {
    handle: "test-3",
    name: "Test 3",
    description: "Test Description 3",
    numEmployees: 300,
    logoUrl: "/logos/logo3.png"
  }
];

const findAllCompanies = jest.fn(() => null); // Learned how to mock functions at https://jestjs.io/docs/mock-functions

it("renders without crashing", () => {
  render(<MemoryRouter><CompanyList companies={ companies } /></MemoryRouter>);
});

it("matches snapshot", () => {
  const { asFragment } = render(
    <MemoryRouter><CompanyList companies={ companies } /></MemoryRouter>
  );
  
  expect(asFragment()).toMatchSnapshot();
});

it("shows list of companies", () => {
  const { queryByTestId } = render(
    <MemoryRouter><CompanyList companies={ companies } /></MemoryRouter>
  );

  for (let company of companies) {
    expect(queryByTestId(company.handle)).toBeInTheDocument();
  }
});

it("updates searchbox when user types", () => {
  const { queryByPlaceholderText, queryByDisplayValue } = render(
    <MemoryRouter><CompanyList companies={ companies } /></MemoryRouter>
  );

  const searchBox = queryByPlaceholderText("Enter search term...");
  
  fireEvent.change(searchBox, { target: { value: "1" } });

  expect(queryByDisplayValue("1")).toBeInTheDocument();
});

it("calls findAllCompanies with no filter when searchbox is empty", () => {
  const { queryByText } = render(
    <MemoryRouter>
      <CompanyList
        companies={ companies }
        findAllCompanies={ findAllCompanies }
      />
    </MemoryRouter>
  );

  const submitButton = queryByText("Submit");
  
  fireEvent.click(submitButton);

  expect(findAllCompanies.mock.calls.length).toEqual(1);
  expect(findAllCompanies.mock.calls[0].length).toEqual(0);
});

it("calls findAllCompanies with filter when searchbox is empty", () => {
  const { queryByPlaceholderText, queryByText } = render(
    <MemoryRouter>
      <CompanyList
        companies={ companies }
        findAllCompanies={ findAllCompanies }
      />
    </MemoryRouter>
  );

  const searchBox = queryByPlaceholderText("Enter search term...");
  
  fireEvent.change(searchBox, { target: { value: "1" } });

  const submitButton = queryByText("Submit");
  
  fireEvent.click(submitButton);

  expect(findAllCompanies.mock.calls.length).toEqual(1);
  expect(findAllCompanies.mock.calls[0][0]).toEqual("1");
});