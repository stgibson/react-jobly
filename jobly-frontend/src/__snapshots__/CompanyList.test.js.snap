// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="CompanyList container"
  >
    <form
      class=""
    >
      <div
        class="form-group row"
      >
        <div
          class="p-0 col-10"
        >
          <input
            class="form-control"
            id="formSearchBox"
            placeholder="Enter search term..."
            type="text"
            value=""
          />
        </div>
        <div
          class="col-2"
        >
          <button
            class="btn btn-primary"
            type="submit"
          >
            Submit
          </button>
        </div>
      </div>
    </form>
    <div
      class="row"
    >
      <div
        class="p-0 col-12"
      >
        <a
          href="/companies/test-1"
        >
          <div
            class="card"
            data-testid="test-1"
          >
            <div
              class="card-body"
            >
              <div
                class="card-title h5"
              >
                <div
                  class="row"
                >
                  <div
                    class="col-10"
                  >
                    Test 1
                  </div>
                  <div
                    class="col-2"
                  >
                    <div
                      data-testid="/logos/logo1.png"
                      style="width: 50px; content: url(/logos/logo1.png);"
                    >
                       
                    </div>
                  </div>
                </div>
              </div>
              <p
                class="card-text"
              >
                Test Description 1
              </p>
            </div>
          </div>
        </a>
        <a
          href="/companies/test-2"
        >
          <div
            class="card"
            data-testid="test-2"
          >
            <div
              class="card-body"
            >
              <div
                class="card-title h5"
              >
                <div
                  class="row"
                >
                  <div
                    class="col-10"
                  >
                    Test 2
                  </div>
                  <div
                    class="col-2"
                  >
                    <div
                      data-testid="/logos/logo2.png"
                      style="width: 50px; content: url(/logos/logo2.png);"
                    >
                       
                    </div>
                  </div>
                </div>
              </div>
              <p
                class="card-text"
              >
                Test Description 2
              </p>
            </div>
          </div>
        </a>
        <a
          href="/companies/test-3"
        >
          <div
            class="card"
            data-testid="test-3"
          >
            <div
              class="card-body"
            >
              <div
                class="card-title h5"
              >
                <div
                  class="row"
                >
                  <div
                    class="col-10"
                  >
                    Test 3
                  </div>
                  <div
                    class="col-2"
                  >
                    <div
                      data-testid="/logos/logo3.png"
                      style="width: 50px; content: url(/logos/logo3.png);"
                    >
                       
                    </div>
                  </div>
                </div>
              </div>
              <p
                class="card-text"
              >
                Test Description 3
              </p>
            </div>
          </div>
        </a>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
